circuit_knitting_toolbox/circuit_cutting/cutting_decomposition.py

```diff
@@ -83,6 +92,132 @@ def partition_circuit_qubits(
+        A tuple containing a dictionary mapping a partition label to a subcircuit
+        and a dictionary mapping a partition label to a list of Pauli observables
+
+    Raises:
+        ValueError: The number of partition labels does not equal the number of qubits in the circuit.
+        ValueError: An input observable acts on a different number of qubits than the input circuit.
+    """
+    if len(partition_labels) != circuit.num_qubits:
+        raise ValueError(
+            f"The number of partition labels ({len(partition_labels)}) must equal the number "
+            f"of qubits in the circuit ({circuit.num_qubits})."
+        )
+    if (
+        observables is not None
+        and len([obs for obs in observables if len(obs) != circuit.num_qubits]) != 0
+    ):
+        raise ValueError(
+            "An input observable acts on a different number of qubits than the input circuit."
+        )
+
+    # Partition the circuit with TwoQubitQPDGates and assign the order via their labels
+    qpd_circuit = partition_circuit_qubits(circuit, partition_labels)
+    i = 0
+    for inst in qpd_circuit.data:
+        if isinstance(inst.operation, TwoQubitQPDGate):
+            inst.operation.label = inst.operation.label + f"_{i}"
+            i += 1
+
+    # Separate the decomposed circuit into its subcircuits
+    qpd_circuit_dx = qpd_circuit.decompose(TwoQubitQPDGate)
+    separated_circs = separate_circuit(qpd_circuit_dx, partition_labels)
+
+    # Decompose the observables, if provided
+    subobservables_by_subsystem = None
+    if observables:
+        subobservables_by_subsystem, _ = decompose_observables(
+            observables, partition_labels
+        )
+
+    return separated_circs.subcircuits, subobservables_by_subsystem
+
+
+def decompose_observables(
+    observables: PauliList, partition_labels: Sequence[str | int]
+) -> tuple[dict[str | int, PauliList], dict[str | int, ObservableCollection]]:
+    """
+    Decompose a list of observables with respect to some qubit partition labels.
+
+    Args:
+        observables: A list of observables to decompose
+        partition_labels: A sequence of partition labels, such that each label
+            corresponds to the qubit in the same index
+
+    Returns:
+        A tuple containing a dictionary mapping a partition to its associated sub-observables
+        and a dictionary mapping a partition to an ``ObservableCollection``
+    """
+    qubits_by_subsystem = defaultdict(list)
+    for i, label in enumerate(partition_labels):
+        qubits_by_subsystem[label].append(i)
+    qubits_by_subsystem = dict(qubits_by_subsystem)  # type: ignore
+
+    subobservables_by_subsystem = {
+        label: observables_restricted_to_subsystem(qubits, observables)
+        for label, qubits in qubits_by_subsystem.items()
+    }
+
+    subsystem_observables = {
+        label: ObservableCollection(subobservables)
+        for label, subobservables in subobservables_by_subsystem.items()
+    }
```

Comment:
I think we should remove this and return only `subobservables_by_subsystem`.  It's needed in two places in the code, and in those places we can just compute it like this.  In fact, in `cutting_reconstruction.py`, a comprehension like this already appears in one of the branches.